drivers/gpio/gpio_tle9104.c

```diff
@@ -0,0 +1,544 @@
+ */
+
+#define DT_DRV_COMPAT infineon_tle9104
+
+#include <errno.h>
+
+#include <zephyr/kernel.h>
+#include <zephyr/device.h>
+#include <zephyr/init.h>
+#include <zephyr/drivers/gpio.h>
+#include <zephyr/drivers/gpio/gpio_utils.h>
+#include <zephyr/drivers/spi.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/sys/byteorder.h>
+
+LOG_MODULE_REGISTER(gpio_tle9104, CONFIG_GPIO_LOG_LEVEL);
+
+/*
+ * The values for the defines below as well as the register definitions were
+ * taken from the datasheet, which can be found at:
+ * https://www.infineon.com/dgdl/Infineon-TLE9104SH-DataSheet-v01_31-EN.pdf?fileId=5546d462766cbe86017676144d76581b
+ */
+#define TLE9104_RESET_DURATION_TIME_IN_US                    10
+#define TLE9104_RESET_DURATION_WAIT_TIME_SAFETY_MARGIN_IN_US 200
+#define TLE9104_RESET_DURATION_WAIT_TIME_IN_US               10
+#define TLE9104_GPIO_COUNT                                   4
+#define TLE9104_INITIALIZATION_TIMEOUT_IN_MS                 1
+#define TLE9104_ICVERSIONID                                  0xB1
+
+#define TLE9104_FRAME_RW_POS                 15
+#define TLE9104_FRAME_PARITY_POS             14
+#define TLE9104_FRAME_FAULTCOMMUNICATION_POS 13
+#define TLE9104_FRAME_FAULTGLOBAL_POS        12
+#define TLE9104_FRAME_ADDRESS_POS            8
+#define TLE9104_FRAME_DATA_POS               0
+
+#define TLE9104_CFG_CWDTIME_LENGTH 2
+#define TLE9104_CFG_CWDTIME_POS    6
+
+#define TLE9104_CTRL_OUT1ONS_BIT       BIT(1)
+#define TLE9104_CTRL_OUT1ONC_BIT       BIT(0)
+#define TLE9104_CFG_OUT1DD_BIT         BIT(0)
+#define TLE9104_GLOBALSTATUS_OUTEN_BIT BIT(7)
+#define TLE9104_GLOBALSTATUS_POR_LATCH BIT(0)
+
+enum tle9104_register {
+	TLE9104REGISTER_CTRL = 0x00,
+	TLE9104REGISTER_CFG = 0x01,
+	TLE9104REGISTER_GLOBALSTATUS = 0x07,
+	TLE9104REGISTER_ICVID = 0x08,
+};
+
+struct tle9104_config {
+	/* gpio_driver_config needs to be first */
+	struct gpio_driver_config common;
+
+	struct spi_dt_spec bus;
+	const struct gpio_dt_spec gpio_reset;
+	const struct gpio_dt_spec gpio_enable;
+	const struct gpio_dt_spec gpio_control[TLE9104_GPIO_COUNT];
+};
+
+struct tle9104_data {
+	/* gpio_driver_data needs to be first */
+	struct gpio_driver_data common;
+	/* each bit is one output channel, bit 0 = OUT1, ... */
+	uint8_t state;
+	/* same as state, just kept for checking what has to be updated */
+	uint8_t previous_state;
+	/* each bit defines if the output channel is configured, see state */
+	uint8_t configured;
+	struct k_mutex lock;
+};
+
+static void tle9104_set_cfg_cwdtime(uint8_t *destination, uint8_t value)
+{
+	uint8_t length = TLE9104_CFG_CWDTIME_LENGTH;
+	uint8_t pos = TLE9104_CFG_CWDTIME_POS;
+
+	*destination &= ~GENMASK(pos + length - 1, pos);
+	*destination |= FIELD_PREP(GENMASK(pos + length - 1, pos), value);
+}
+
+static int tle9104_calculate_parity(uint16_t value)
+{
+	int parity = 1;
+
+	if ((value & BIT(TLE9104_FRAME_RW_POS)) != 0) {
+		parity++;
+	}
+
+	for (size_t i = 0; i < TLE9104_FRAME_PARITY_POS; ++i) {
```

Comment:
Hint:
You could use popcount here

Reply:
Learned something new today, thanks.
done

Reply:
damn, cool!